source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs

```diff
@@ -360,54 +406,111 @@ public async Task DequeueAsync_WillContinueWaitingUntilItemIsQueued(bool async)
             dequeuedRequest.Should().Be(request);
 
             // Apply a response so we can prove this counts as taking a message.
-            sut.ApplyResponse(expectedResponse, request.Destination);
+            await sut.ApplyResponse(expectedResponse, request.Destination);
             var response = await queueAndWaitTask;
             response.Should().Be(expectedResponse);
         }
+        
+        [Test]
+        [SyncAndAsync]
+        public async Task DequeueAsync_WithMultipleDequeueCallsWaiting_WhenSingleRequestIsQueued_ThenOnlyOneCallersReceivesRequest(SyncOrAsync syncOrAsync)
+        {
+            // Arrange
+            const string endpoint = "poll://endpoint001";
+
+            var sut = new PendingRequestQueueBuilder().WithAsync(syncOrAsync).WithEndpoint(endpoint).Build();
+            var request = new RequestMessageBuilder(endpoint).Build();
+            var expectedResponse = ResponseMessageBuilder.FromRequest(request).Build();
+            
+            var dequeueTasks = Enumerable.Range(0, 30)
+                .Select(_ => sut.DequeueAsync(CancellationToken))
+                .ToArray();
+
+            await Task.Delay(1000, CancellationToken);
+
+            // Act
+            var queueAndWaitTask = StartQueueAndWait(sut, request, CancellationToken);
+
+            // Assert
+            await Task.WhenAll(dequeueTasks);
+
+            await sut.ApplyResponse(expectedResponse, request.Destination);
+            await queueAndWaitTask;
+
+            var singleDequeuedRequest = dequeueTasks.Should().ContainSingle(t => t.Result != null).Subject.Result;
+            singleDequeuedRequest.Should().Be(request);
+
+        }
+
+        [Test]
+        [SyncAndAsync]
+        public async Task ApplyResponse_AfterRequestHasBeenCollected_AndWaitingHasBeenCancelled_ResponseIsIgnored(SyncOrAsync syncOrAsync)
```

Comment:
There was no real way to test the memory leak fix. This was the best usecase I could think of that came close.